### DISCUSSION TITLE
How do http get requests better define query string parameters in Pydantic Model

### DISCUSSION THREAD

**Discussion by HuiDBK** [fastapi/fastapi#10016] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI, Query, Depends
from pydantic import BaseModel, Field
from typing import Optional

app = FastAPI()


class ListPageReqModel(BaseModel):
    """分页请求模型"""

    cur_page: int = Field(default=1, gt=0, description="当前页码")
    page_num: int = Field(default=20, gt=0, le=1000, description="每页数量")


class ListPageQueryModel(BaseModel):
    """分页请求模型"""

    cur_page: int = Field(Query(default=1, gt=0, description="当前页码"))
    page_num: int = Field(Query(default=20, gt=0, le=1000, description="每页数量"))


class UserQueryIn(ListPageQueryModel):
    name: Optional[str] = Field(Query(default=None, description="姓名模糊查询"))
    age: Optional[int] = Field(Query(default=None, description="年龄筛选"))


class UserReqIn(ListPageReqModel):
    name: Optional[str] = Field(default=None, description="姓名模糊查询")
    age: Optional[int] = Field(default=None, description="年龄筛选")


@app.get("/api/v1/users")
async def users(
        name: Optional[str] = Query(default=None, description="姓名模糊查询"),
        age: Optional[int] = Query(default=None, description="年龄筛选"),
        page_params: ListPageReqModel = Depends(ListPageReqModel)

):
    return {"users": [{"name": "hui"}]}


@app.get("/api/v2/users")
async def users(req_model: UserReqIn = Depends(UserReqIn)):
    return {"users": [{"name": "hui"}]}


@app.get("/api/v3/users")
async def users(req_model: UserQueryIn = Depends(UserQueryIn)):
    return {"users": [{"name": "hui"}]}


# I wish it were that simple
# AssertionError: Param: req_model can only be a request body, using Body()
# @app.get("/api/v4/users")
# async def users(req_model: UserReqIn = Query(UserReqIn)):
#     return {"users": [{"name": "hui"}]}


# This is easiest, but req_model is not aware of the type and is syntactically incorrect
# TypeError: Annotated[t, ...]: t must be a type. Got Query(<class '__main__.UserReqIn'>).
# @app.get("/api/v5/users")
# async def users(req_model: Query(UserReqIn)):
#     return {"users": [{"name": "hui"}]}


if __name__ == '__main__':
    uvicorn.run(app)
```

### Description

In Pydantic Model, it is inconvenient to rely on Depends and Query to assist in constructing query string parameters. If Depends alone can be used, but the OpenAPI document generated by /docs does not have a description of the corresponding field, Depedns and Query can be used together to achieve this.
It would be so much easier if we could just make Query(pydantic_query_model) supported

<img width="663" alt="image" src="https://github.com/tiangolo/fastapi/assets/45389728/69c8377c-f127-4461-8d03-8b88179a54c2">

<img width="916" alt="image" src="https://github.com/tiangolo/fastapi/assets/45389728/1f7aec0b-1e08-4840-8788-5af1f056466e">

<img width="727" alt="image" src="https://github.com/tiangolo/fastapi/assets/45389728/c965fa73-3a19-4699-9e16-c90e7d4b04e5">

### Operating System

macOS

### FastAPI Version

0.100.1

### Pydantic Version

2.1.1

### Python Version

3.10.12

**Comment by Angry-Chick** [comment]
I also encountered the same problem. thank you for the third temporary solution and hope this issue can be improved in the future

**Comment by AlexanderPodorov** [comment]
You can use plain dataclasses or pydantic dataclasses to achieve that. See classes as dependencies section in the docs.

**Comment by YuriiMotov** [comment, marked as answer]
Since 0.115.0 FastAPI supports Query models: https://fastapi.tiangolo.com/tutorial/query-param-models/
